text/0000-guaranteed-slice-repr.md

```diff
@@ -0,0 +1,182 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Slices are represented with a pointer and length pair. Their in-memory layout is
+the same as a `#[repr(C)]` struct like the following:
+
+```rust
+#[repr(C)]
+struct Slice<T> {
+    data: *const T,
+    len: usize,
+}
+```
+
+The precise ABI of slices is not guaranteed, so `&[T]` may not be passed by-value
+or returned by-value from an `extern "C" fn`.
+
+The validity requirements for the in-memory slice representation are the same
+as [those documented on `std::slice::from_raw_parts`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html).
+Namely:
+
+* `data` must be non-null, valid for reads for `len * mem::size_of::<T>()` many bytes,
+  and it must be properly aligned. This means in particular:
+
+    * The entire memory range of this slice must be contained within a single allocated object!
+      Slices can never span across multiple allocated objects.
+    * `data` must be non-null and aligned even for zero-length slices or slices of ZSTs. One
+      reason for this is that enum layout optimizations may rely on references
+      (including slices of any length) being aligned and non-null to distinguish
+      them from other data. You can obtain a pointer that is usable as `data`
+      for zero-length slices using [`NonNull::dangling()`].
+
+* `data` must point to `len` consecutive properly initialized values of type `T`.
+
+* The total size `len * mem::size_of::<T>()` of the slice must be no larger than `isize::MAX`,
+  and adding that size to `data` must not "wrap around" the address space.
+  See the safety documentation of [`pointer::offset`].
+
+## `str`
+
+The layout of `&str` is the same as that of `&[u8]`, and the layout of
+&mut str` is the same as that of `&mut [u8]`. More generally, `str` behaves like
+`#[repr(transparent)] struct str([u8]);`. Safe Rust functions may assume that
+`str` holds valid UTF8, but [it is not immediate undefined-behavior to store
+non-UTF8 data in `str`](https://doc.rust-lang.org/std/primitive.str.html#invariant).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Zero-sized types
+
+One could imagine representing `&[T]` as only `len` for zero-sized `T`.
+This proposal would preclude that choice in favor of a standard representation
+for slices regardless of the underlying type.
+
+Alternatively, we could choose to guarantee that the data pointer is present if
+and only if `size_of::<T> != 0`. This has the possibility of breaking exising
+code which smuggles pointers through the `data` value in `from_raw_parts` /
+`into_raw_parts`.
+
+## Uninhabited types
+
+Similarly, we could be *extra* tricky and make `&[!]` or other `&[Uninhabited]`
+types into a ZST since the slice can only ever be length zero. This may offer
+modest performance benefits for highly generic code which happens to create
+empty slices of uninhabited types, but this is unlikely to be worth the
+cost of maintaining a special case.
```

Comment:
We could also make them length 1 by omitting the length but keeping the pointer. (I don’t necessarily think we should do this, just raising the possibility. Either way, raw pointers should keep the length field, for compatibility.)